pithos/pithos.py

```diff
@@ -1051,24 +1054,80 @@ def show_about(self, version):
-            self.on_explicit_content_filter_checkbox()
-            if get_account_password(email) != self.last_pass:
+            if self.last_password != new_password or self.last_email != new_email:
+                #full stop, reset and reconnect on password and/or email change.
+                self.stop()
+                self.waiting_for_playlist = False
+                self.current_song_index = None
+                self.start_new_playlist = False
+                self.current_station = None
+                self.current_station_id = None
+                self.have_stations = False
+                self.playcount = 0
+                self.songs_model.clear()
                 self.pandora_connect()
-        else:
-            if not email or not get_account_password(email):
-                self.quit()
-        self.last_pass = ''
+
+            elif (self.last_proxy != new_proxy
+                  or self.last_control_proxy != new_control_proxy
+                  or self.last_control_proxy_pac != new_control_proxy_pac
+                  or self.last_one_checkbox != new_one_checkbox):
+                #just a reconnect on proxy or pandora_one change.
+                self.pandora_connect()
+            self.on_explicit_content_filter_checkbox()
+
+        if response == Gtk.ResponseType.CANCEL:
+            #password doesn't properly revert on cancel.
+            self.prefs_dlg.password_entry.set_text(self.last_password)
+            #volume should not revert on cancel.
+            self.settings.set_value('volume', self.last_volume)             
+            if not self.prefs_dlg.email_entry.get_text() or not self.prefs_dlg.password_entry.get_text():
+                #error dialog if password and email are left blank on prefs cancel. Should only ever happen on 1st run.
+                self.error_dialog(_("Pandora Account Credentials Required"), None,
+                                  submsg=_("Please enter your email and password."))
 
     def show_preferences(self):
         """preferences - display the preferences window for pithos """
         self.sync_explicit_content_filter_setting()
-        self.last_pass = get_account_password(self.settings['email'])
+
+        self.last_email = self.prefs_dlg.email_entry.get_text()
+        self.last_password = self.prefs_dlg.password_entry.get_text()
+        self.last_proxy = self.prefs_dlg.proxy_entry.get_text()
+        self.last_control_proxy = self.prefs_dlg.control_proxy_entry.get_text()
+        self.last_control_proxy_pac = self.prefs_dlg.control_proxy_pac_entry.get_text()
+        self.last_one_checkbox = self.prefs_dlg.pandora_one_checkbutton.get_active()
+        self.last_volume = self.settings.get_value('volume')
+
         self.settings.delay() # Dialog will apply
         self.prefs_dlg.show()
 
+    def on_password_entry_grab_focus(self, *ignore):
+        #clear out the password_entry and uncheck 
+        #pandora_one_checkbutton when it's clicked into.
```

Comment:
Think of all the elderly who will have a heart attack when they lose their password!

(but seriously I don't like it and have never seen such a thing in my life)